api/v1alpha1/es_types.go

```diff
@@ -0,0 +1,195 @@
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster
+
+// ExternalSecrets describes configuration and information about the managed external-secrets
+// deployment. The name must be `cluster` to make ExternalSecrets a singleton that is, to
+// allow only one instance of ExternalSecrets per cluster.
+//
+// When an ExternalSecrets is created, a new deployment is created which manages the
+// external-secrets and keeps it in the desired state.
+//
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'cluster'",message="ExternalSecrets is a singleton, .metadata.name must be 'cluster'"
+// +operator-sdk:csv:customresourcedefinitions:displayName="ExternalSecrets"
+type ExternalSecrets struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the ExternalSecrets.
+	// +kubebuilder:validation:Required
+	Spec ExternalSecretsSpec `json:"spec,omitempty"`
+
+	// status is the most recently observed status of the ExternalSecrets.
+	Status ExternalSecretsStatus `json:"status,omitempty"`
+}
+
+// ExternalSecretsSpec is the specification of the desired behavior of the ExternalSecrets.
+type ExternalSecretsSpec struct {
+	// externalSecretsConfig is for configuring the external-secrets behavior.
+	// +kubebuilder:validation:Optional
+	ExternalSecretsConfig *ExternalSecretsConfig `json:"externalSecretsConfig,omitempty"`
+
+	// controllerConfig is for configuring the controller for setting up
+	// defaults to enable external-secrets.
+	// +kubebuilder:validation:Optional
+	ControllerConfig *ControllerConfig `json:"controllerConfig,omitempty"`
+}
+
+// ExternalSecretsStatus is the most recently observed status of the ExternalSecrets.
+type ExternalSecretsStatus struct {
+	// conditions holds information of the current state of the external-secrets deployment.
+	ConditionalStatus `json:",inline,omitempty"`
+}
+
+// ExternalSecretsConfig is for configuring the external-secrets behavior.
+type ExternalSecretsConfig struct {
+	// logLevel supports value range as per [kubernetes logging guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use).
+	// +kubebuilder:default:=1
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=5
+	// +kubebuilder:validation:Optional
+	LogLevel int32 `json:"logLevel,omitempty"`
+
+	// bitwardenSecretManagerProvider is for enabling the bitwarden secrets manager provider and
+	// for setting up the additional service required for connecting with the bitwarden server.
+	// +kubebuilder:validation:Optional
+	BitwardenSecretManagerProvider *BitwardenSecretManagerProvider `json:"bitwardenSecretManagerProvider,omitempty"`
```

Comment:
I'm willing to understand why we require the Bitwarden provider config in here, while do not require it for the other providers, could you help shed some light on it?

Reply:
Bitwarden provider is a feature supported by `external-secrets` which is configurable and the same is extended here for user to configure. 
Ref: https://external-secrets.io/latest/provider/bitwarden-secrets-manager/

Reply:
Ah, so what I understood is that external-secrets [controller] makes use of an optional Bitwarden SDK server, we're trying to provide for customizing that aspect in our offering through this config field (i.e. as a functional alternative to enabling it through optional flag in the upstream helm chart).

Also, importantly no other providers require similar config (which is what caused my initial confusion, thanks for clarifying).